galasa-parent/dev.galasa.framework/src/main/java/dev/galasa/framework/spi/creds/FrameworkEncryptionService.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright contributors to the Galasa project
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package dev.galasa.framework.spi.creds;
+
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Paths;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+
+import javax.crypto.Cipher;
+import javax.crypto.spec.GCMParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+
+import org.yaml.snakeyaml.LoaderOptions;
+import org.yaml.snakeyaml.Yaml;
+import org.yaml.snakeyaml.constructor.Constructor;
+
+import dev.galasa.framework.FileSystem;
+import dev.galasa.framework.IFileSystem;
+import dev.galasa.framework.spi.Environment;
+import dev.galasa.framework.spi.SystemEnvironment;
+
+public class FrameworkEncryptionService implements IEncryptionService {
+
+    public static final String ENCRYPTION_KEYS_PATH_ENV = "GALASA_ENCRYPTION_KEYS_PATH";
+
+    private static final String KEY_ALGORITHM = "AES";
+    private static final String ENCRYPTION_ALGORITHM = "AES/GCM/NoPadding";
+
+    private static final int GCM_AUTH_TAG_LENGTH_BITS = 128;
+    private static final int GCM_IV_BYTES_LENGTH = 12;
+
+    private SecretKeySpec encryptionKey;
+    private List<SecretKeySpec> decryptionKeys = new ArrayList<>();
+
+    private Environment environment = new SystemEnvironment();
+    private IFileSystem fileSystem = new FileSystem();
+
+    public FrameworkEncryptionService(SecretKeySpec encryptionKey) throws CredentialsException {
+        this(encryptionKey, new FileSystem(), new SystemEnvironment());
+    }
+
+    public FrameworkEncryptionService(SecretKeySpec encryptionKey, IFileSystem fileSystem, Environment environment) throws CredentialsException {
+        this.fileSystem = fileSystem;
+        this.environment = environment;
+        this.encryptionKey = encryptionKey;
+
+        if (encryptionKey == null) {
+            EncryptionKeys encryptionKeysYaml = parseEncryptionKeysFile();
+            if (encryptionKeysYaml != null) {
+                this.decryptionKeys = loadDecryptionKeys(encryptionKeysYaml);
+                this.encryptionKey = loadPrimaryEncryptionKey(encryptionKeysYaml);
+            }
+        } else {
+            this.decryptionKeys = new ArrayList<>();
+            this.decryptionKeys.add(encryptionKey);
+        }
+    }
+
+    public FrameworkEncryptionService(IFileSystem fileSystem, Environment environment) throws CredentialsException {
+        this(null, fileSystem, environment);
+    }
+
+    private EncryptionKeys parseEncryptionKeysFile() throws CredentialsException {
```

Comment:
I think it would help to have an example of a yaml file to show what it's trying to parse

Reply:
Might be worthwhile separating out an EncryptionKeys class which can parse a yaml into itself ?